File: lintcode/number-of-islands.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/number-of-islands/

class Solution:
    def dfs(self, grid, x, y):
        if x >= len(grid) or y >= len(grid[0]) or x < 0 or y < 0:
            return
        if grid[x][y] != 1:
            return
        grid[x][y] = -1
        self.dfs(grid, x + 1, y)
        self.dfs(grid, x, y + 1)
        self.dfs(grid, x - 1, y)
        self.dfs(grid, x, y - 1)

    # @param {boolean[][]} grid a boolean 2D matrix
    # @return {int} an integer
    def numIslands(self, grid):
        # Write your code here
        if len(grid) == 0 or len(grid[0]) == 0:
            return 0

        count = 0
        for i in xrange(len(grid)):
            for j in xrange(len(grid[0])):
                if grid[i][j] == 1:
                    count += 1
                    self.dfs(grid, i, j)

        return count

if __name__ == '__main__':
    s = Solution()
    print s.numIslands([
        [1, 1, 0, 0, 0],
        [0, 1, 0, 0, 1],
        [0, 0, 0, 1, 1],
        [0, 0, 0, 0, 0],
        [0, 0, 0, 0, 1]
    ])

File: leetcode/number-of-islands.cc
// Question: https://leetcode.com/problems/number-of-islands/

#include "headers.h"

class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int ans = 0;
        for (int i = 0; i < grid.size(); ++i) {
            for (int j = 0; j < grid[0].size(); ++j) {
                if (grid[i][j] == '1') {
                    dfs(grid, i, j);
                    ++ans;
                }
            }
        }
        return ans;
    }

private:
    void dfs(vector<vector<char>>& grid, int x, int y) {
        if (x >= grid.size() || y >= grid[0].size() || x < 0 || y < 0) {
            return;
        }
        if (grid[x][y] == '1') {
            grid[x][y] = '0';
            dfs(grid, x+1, y);
            dfs(grid, x-1, y);
            dfs(grid, x, y+1);
            dfs(grid, x, y-1);
        }
    }
};

int main(int argc, char** argv) {
    Solution s;
    {
        vector<vector<char>> grid;
        grid.push_back(vector<char>{'1','1','1','1','0'});
        grid.push_back(vector<char>{'1','1','0','1','0'});
        grid.push_back(vector<char>{'1','1','0','0','0'});
        grid.push_back(vector<char>{'0','0','0','0','0'});
        cout << s.numIslands(grid) << endl;
    }
    {
        vector<vector<char>> grid;
        grid.push_back(vector<char>{'0'});
        cout << s.numIslands(grid) << endl;
    }
}
